aws/rust-runtime/aws-runtime/src/env_config.rs

```diff
@@ -0,0 +1,548 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use crate::env_config::property::PropertiesKey;
+use crate::env_config::section::EnvConfigSections;
+use aws_types::os_shim_internal::Env;
+use aws_types::service_config::ServiceConfigKey;
+use std::borrow::Cow;
+use std::error::Error;
+use std::fmt;
+
+pub mod error;
+pub mod file;
+mod normalize;
+pub mod parse;
+pub mod property;
+pub mod section;
+pub mod source;
+
+/// Given a key, access to the environment, and a validator, return a config value if one was set.
+pub fn get_service_env_config<'a, T, E>(
+    key: ServiceConfigKey<'a>,
+    env: &'a Env,
+    shared_config_sections: Option<&'a EnvConfigSections>,
+    validator: impl Fn(&str) -> Result<T, E>,
+) -> Result<Option<T>, EnvConfigError<E>>
+where
+    E: Error + Send + Sync + 'static,
+{
+    EnvConfigValue::default()
+        .env(key.env())
+        .profile(key.profile())
+        .service_id(key.service_id())
+        .validate(env, shared_config_sections, validator)
+}
+
+#[derive(Debug)]
+enum Location<'a> {
+    Environment,
+    Profile { name: Cow<'a, str> },
+}
+
+impl<'a> fmt::Display for Location<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Location::Environment => write!(f, "environment variable"),
+            Location::Profile { name } => write!(f, "profile (`{name}`)"),
+        }
+    }
+}
+
+#[derive(Debug)]
+enum Scope<'a> {
+    Global,
+    Service { service_id: Cow<'a, str> },
+}
+
+impl<'a> fmt::Display for Scope<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Scope::Global => write!(f, "global"),
+            Scope::Service { service_id } => write!(f, "service-specific (`{service_id}`)"),
+        }
+    }
+}
+
+/// The source that env config was derived from.
+///
+/// Includes:
+///
+/// - Whether some config came from a config file or an env var.
+/// - The key used to identify the config value.
+///
+/// Only used when displaying config-extraction errors.
+#[derive(Debug)]
+pub struct EnvConfigSource<'a> {
+    key: Cow<'a, str>,
+    location: Location<'a>,
+    source: Scope<'a>,
```

Comment:
nit: just want to check whether `source` is intended as opposed to `scope` (minor thing since it's private).

Reply:
good catch